tsl/src/remote/dist_ddl.c

```diff
@@ -390,9 +420,23 @@ dist_ddl_preprocess(ProcessUtilityArgs *args)
 			break;
 
 		case T_CreateTrigStmt:
+		{
+			CreateTrigStmt *stmt = castNode(CreateTrigStmt, args->parsetree);
+
+			if (stmt->funcname != NIL)
+			{
+				Oid funcargtypes[1];
+				Oid funcid = LookupFuncName(stmt->funcname,
+											0 /* nargs */,
+											funcargtypes /* passing NULL is not allowed */,
+											false /* missing ok */);
+				Datum datum = DirectFunctionCall1(pg_get_functiondef, ObjectIdGetDatum(funcid));
+
+				dist_ddl_add_remote_command(TextDatumGetCString(datum));
```

Comment:
So it basically will recreate the trigger function on each trigger creation, right? I am thinking if there are any downsides of it. What if user previously distributed this function on data nodes, I guess previous method will be no longer recommended (like using distrubted_exec for instance).

Reply:
Yes, if multiple triggers use the same function, the function will be recreated. The only odd case I can think of is when a creates the trigger function on a few data nodes, then updates the function on the AN, then attaches another data node. Only the new data node will get the new version of trigger function unless the user explicitly updates it on other data nodes.

Still, that problem exists even when users have to manually create the function. Although I guess then it is explicit that it is up to them to fix it.

Reply:
I am not entirely convinced that it is a good idea to unconditionally replace the function on the data node, but for now I think we can accept that data nodes is not something that you should mess with so if you create a function here you have to accept that it might be overwritten.

Reply:
@mkindahl the principle we have decided on is that users interact with the AN and not DNs (if they do, they're on their own). So, distributed objects, like the trigger and function here, should exist in identical form on all nodes. Admittedly, this isn't yet perfect, given that users can still change the function on the AN without having it updated on DNs. But, recreating the function on DNs based on the ANs function doesn't violate the principle.

Reply:
> @mkindahl the principle we have decided on is that users interact with the AN and not DNs (if they do, they're on their own). So, distributed objects, like the trigger and function here, should exist in identical form on all nodes. Admittedly, this isn't yet perfect, given that users can still change the function on the AN without having it updated on DNs. But, recreating the function on DNs based on the ANs function doesn't violate the principle.

Yes? Wasn't that what I said?

Reply:
Since you weren't convinced, I wanted to provide additional context as to why this seems to not violate established principles.